extensions/ql-vscode/src/test-manager.ts

```diff
@@ -0,0 +1,373 @@
+    op(testItem);
+  }
+}
+
+/**
+ * Implementation of `BaseLogger` that logs to the output of a `TestRun`.
+ */
+class TestRunLogger implements BaseLogger {
+  public constructor(private readonly testRun: TestRun) {}
+
+  public async log(message: string, options?: LogOptions): Promise<void> {
+    // "\r\n" because that's what the test terminal wants.
+    const lineEnding = options?.trailingNewline === false ? "" : "\r\n";
+    this.testRun.appendOutput(message + lineEnding);
+  }
+}
+
+/**
+ * Handles test discovery for a specific workspace folder, and reports back to `TestManager`.
+ */
+class WorkspaceFolderHandler extends DisposableObject {
+  private readonly testDiscovery: QLTestDiscovery;
+
+  public constructor(
+    private readonly workspaceFolder: WorkspaceFolder,
+    private readonly testUI: TestManager,
+    cliServer: CodeQLCliServer,
+  ) {
+    super();
+
+    this.testDiscovery = new QLTestDiscovery(workspaceFolder, cliServer);
+    this.push(
+      this.testDiscovery.onDidChangeTests(this.handleDidChangeTests, this),
+    );
+    this.testDiscovery.refresh();
+  }
+
+  private handleDidChangeTests(): void {
+    const testDirectory = this.testDiscovery.testDirectory;
+
+    this.testUI.updateTestsForWorkspaceFolder(
+      this.workspaceFolder,
+      testDirectory,
+    );
+  }
+}
+
+/**
+ * Service that populates the VS Code "Test Explorer" panel for CodeQL, and handles running and
+ * debugging of tests.
+ */
+export class TestManager extends TestManagerBase {
+  /**
+   * Maps from each workspace folder being tracked to the `WorkspaceFolderHandler` responsible for
+   * tracking it.
+   */
+  private readonly workspaceFolderHandlers = new Map<
+    WorkspaceFolder,
+    WorkspaceFolderHandler
+  >();
+
+  public constructor(
+    app: App,
+    private readonly testRunner: TestRunner,
+    private readonly cliServer: CodeQLCliServer,
+    // Having this as a parameter with a default value makes passing in a mock easier.
+    private readonly testController: TestController = tests.createTestController(
+      "codeql",
+      "CodeQL Tests",
+    ),
+  ) {
+    super(app);
+
+    this.testController.createRunProfile(
+      "Run",
+      TestRunProfileKind.Run,
+      this.run.bind(this),
+    );
+
+    // Start by tracking whatever folders are currently in the workspace.
+    this.startTrackingWorkspaceFolders(workspace.workspaceFolders ?? []);
+
+    // Listen for changes to the set of workspace folders.
+    workspace.onDidChangeWorkspaceFolders(
+      this.handleDidChangeWorkspaceFolders,
+      this,
+    );
+  }
+
+  public dispose(): void {
+    this.workspaceFolderHandlers.clear(); // These will be disposed in the `super.dispose()` call.
+    super.dispose();
+  }
+
+  protected getTestPath(node: TestItem): string {
+    if (node.uri === undefined || node.uri.scheme !== "file") {
+      throw new Error("Selected test is not a CodeQL test.");
+    }
+    return node.uri!.fsPath;
```

Comment:
TypeScript should recognize that `node.uri` is now not undefined:

```suggestion
    return node.uri.fsPath;
```

Reply:
Fixed.